soc/arm/nxp_s32/common/power_soc.c

```diff
@@ -15,49 +15,23 @@
 #ifdef CONFIG_REBOOT
 BUILD_ASSERT(POWER_IP_PERFORM_RESET_API == STD_ON, "Power Reset API must be enabled");
 
-/*
- * Overrides default weak implementation of system reboot.
- *
- * SYS_REBOOT_COLD (Destructive Reset):
- * - Leads most parts of the chip, except a few modules, to reset. SRAM content
- *   is lost after this reset event.
- * - Flash is always reset, so an updated value of the option bits is reloaded
- *   in volatile registers outside of the Flash array.
- * - Trimming is lost.
- * - STCU is reset and configured BISTs are executed.
- *
- * SYS_REBOOT_WARM (Functional Reset):
- * - Leads all the communication peripherals and cores to reset. The communication
- *   protocols' sanity is not guaranteed and they are assumed to be reinitialized
- *   after reset. The SRAM content, and the functionality of certain modules, is
- *   preserved across functional reset.
- * - The volatile registers are not reset; in case of a reset event, the
- *   trimming is maintained.
- * - No BISTs are executed after functional reset.
- */
-void sys_arch_reboot(int type)
+void z_sys_reboot(enum sys_reboot_type type)
 {
-	Power_Ip_MC_RGM_ConfigType mc_rgm_cfg = { 0 };
+	ARG_UNUSED(type);
 
-	const Power_Ip_HwIPsConfigType power_cfg = {
-		.McRgmConfigPtr = (const Power_Ip_MC_RGM_ConfigType *)&mc_rgm_cfg,
-		.PMCConfigPtr = NULL
+	Power_Ip_MC_RGM_ConfigType mc_rgm_cfg = {
+		.ResetType = MCU_FUNC_RESET,
```

Comment:
This target had different COLD/WARM implementations, my understanding is that you are dropping WARM usage across the tree, so why here you ignoring the given parameter and arbitrary using the WARM path ? maybe a typo ?

COLD was `mc_rgm_cfg.ResetType = MCU_DEST_RESET;` 
WARM `mc_rgm_cfg.ResetType = MCU_FUNC_RESET;`

Reply:
this is done on purpose. `sys_reboot()` is not expected to perform destructive actions, maybe some new APIs should be introduced https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/d4b0d71107f72effbcce02cd66ada2fa01a0b6b8

If we allow any kind of behavior for sys_reboot() there's no way to write portable code.

Reply:
Looking the rest of the changes it seems it should be:
```
if (typ == SYS_REBOOT_WARM) {
		mc_rgm_cfg.ResetType = MCU_FUNC_RESET;
} else {
		mc_rgm_cfg.ResetType = MCU_DEST_RESET;
}

Power_Ip_PerformReset(&power_cfg);

for (;;) {
		/* wait for reset */
}
```

Reply:
> this is done on purpose. `sys_reboot()` is not expected to perform destructive actions, maybe some new APIs should be introduced [d4b0d71](https://github.com/zephyrproject-rtos/zephyr/commit/d4b0d71107f72effbcce02cd66ada2fa01a0b6b8)
> 
> If we allow any kind of behavior for sys_reboot() there's no way to write portable code.

I am pretty confuse now, isn't cold reboot something that can wipe out your ram, cut the power, ... basically any context is lost ? This target was "supposedly" implementing it. If that is not the case, what it is cold and warm boot for us ? I honestly don't want an infinite discussion, these changes are good, what I don't understand is why this target that implements both WARM / COLD reboots  have the COLD path dropped ... Is that because of other changes in the cold replacing the WARM with COLD ?

If that is the case, we may re-evaluate why having two types of reboot

Reply:
@gmarull I'm okay with the change if `sys_reboot` it's not supposed to perform destructive actions. At the time I did this, it was not clear to me what is the expectation of cold/warm reset, hence the description I've included in this function.

Reply:
Thanks for the review @manuargue. Do you know under which circumstances the destructive reset is used/preferred?

Reply:
> Thanks for the review @manuargue. Do you know under which circumstances the destructive reset is used/preferred?

@gmarull In general for the S32 devices, the request sequence goes like `POR -> destructive reset sequence -> functional reset sequence -> chip out of reset -> perform boot flow`.
Destructive reset (either coming from POR or from an event that triggered it) will reset everything except the POR and debug logic, so this is the closest to POR.
Functional reset, will not reset all modules (e.g. RTC, PMC, fault detections) and SRAM content is preserved.

Reply:
With the destructive reset, that would cause some issues with some zephyr systems e.g. retained RAM. I agree this should be standardised as to what a reset does and preserves.

Reply:
@manuargue so, which reset should be the default in this case? is destructive always preferred or not?

Reply:
+1 for standardization. In general, reset is platform dependent, but we could define what is expected to be reset/preserved for each of the types of reset defined by Zephyr. Otherwise we leave the door open to interpretations and inconsistent behaviors between SoCs.

Reply:
I think it'll be difficult to standardize the reboot modes because there seems to be a wide number of variants in the field. I'd tend to leave warm/cold options, but let implementations decide what do they mean exactly. We can ask for certain guarantees, e.g. warm resets less than cold, it's faster, etc. But I don't think we can expect to write portable code that uses `SYS_REBOOT_COLD/WARM`.

This is the reason `SYS_REBOOT_DEFAULT` https://github.com/zephyrproject-rtos/zephyr/pull/60920/commits/8ed86746da92684b9723068a21ff98fa14962486 has been introduced. If some portable code has no clue on which reset mode to choose, it should provide `SYS_REBOOT_DEFAULT` so that the system decides the _most suitable_ one if not specified.

Also, for reference, it looks like Linux does allow setting distinct reset modes only via Kernel boot parameters https://elixir.bootlin.com/linux/latest/source/kernel/reboot.c#L1068. So it seems like it's something that one consciously decides when booting the Kernel, and so likely knowing the implications of reset modes. `SYS_REBOOT_DEFAULT` goes in line with this approach.

Btw, @manuargue can you review your -1?